torch/csrc/jit/tensorexpr/loopnest.h

```diff
@@ -222,6 +222,39 @@ class TORCH_API LoopNest {
       const std::vector<For*>& loops,
       const std::vector<size_t>& permutation);
 
+  // Tile takes a 2d domain (x, y) and split it into small rectangular blocks
+  // each with shape (x_factor, y_factor). The traversal over the domain turns
+  // into an outter iteration over the blocks and an inner traversal over all
+  // points in the block.
+  // Note that if x dim % x_factor or y dim % y_factor does not equal to 0, the
+  // loop body will be guarded by boundary checks.
```

Comment:
Did you consider splitting such loops instead of adding guards? Does guards give a significant advantage here?

I believe the implementation would be lot simpler if we split the loops and not have guards?

Reply:
I tried to generate simpler stmts for later transformations like `reorder`. Probably we should consider leaving the decision of `splitWithTail` or `splitWithMask` to when we are ready for codegen (the lowering pass)? Otherwise, dealing with a large compound loop stmt after `splitWithTail`s is so painful. @ZolotukhinM

Reply:
Actually, using guards (with `splitWithMask`) would be generating a more complicated code. I don't understand why `splitWithTail` will result in that. Can you show an example?

Reply:
I mocked up a tile in python with `split_with_tail` and it seems working:
```python
import torch

class kernel_arena_scope(object):
    def __enter__(self):
        self.scope = torch._C._te.KernelScope()

    def __exit__(self, typ, val, traceback):
        self.scope = None

def tile(loopnest, loop_x, loop_y, factor_x, factor_y):
    enc_loops = loopnest.get_enclosing_loopnest(loop_y)
    assert(enc_loops[-1] == loop_x)

    yo, yi, yt = loopnest.split_with_tail(loop_y, factor_y)
    print("split y", loopnest)

    loop_x = loopnest.distribute_loop_over_inner_loops(loop_x)[0]
    print("distribute x", loopnest)

    xo, xi, xt = loopnest.split_with_tail(loop_x, factor_x)
    print("split x", loopnest)

    # We need to re-identify yo and yi because after split(x) we have two
    # versions of each: one in the x_inner and another in x_tail
    # But since we explicitly distributed x we know that yo is simply the first
    # and the only stmt in xi. Similarly, yi is the first and only stmt in yo.
    yo = xi.body().stmts()[0]
    yi = yo.body().stmts()[0]
    loopnest.reorder([xo, xi, yo, yi], [0, 2, 1, 3])
    print("reorder", loopnest)

    loopnest.simplify()
    print("simplify", loopnest)

with kernel_arena_scope():
    f32 = torch._C._te.Dtype.Float
    i32 = torch._C._te.Dtype.Int
    M = torch._C._te.VarHandle("M", i32)
    N = torch._C._te.VarHandle("N", i32)

    X = torch._C._te.Placeholder('X', f32, [M, N])
    Y = torch._C._te.Placeholder('Y', f32, [M, N])

    dim_args = [torch._C._te.DimArg(*args) for args in [(M, 'x'), (N, 'y')]]

    def compute(m, n):
        return X.load([m, n]) + Y.load([m, n])
    Z = torch._C._te.Compute('Z', dim_args, compute)
    loopnest = torch._C._te.LoopNest([Z])
    loopnest.simplify()
    print("original", loopnest)
    loops = loopnest.get_loops_for(Z)
    tile(loopnest, loops[0], loops[1], 8, 8)
    print("after tile", loopnest)
```

We still need to re-identify once, but I think it's not a very big deal (also, maybe we actually can avoid it somehow).

Reply:
> # We need to re-identify yo and yi because after split(x) we have two
> # versions of each: one in the x_inner and another in x_tail

If we get rid of the `clone` in `splitWithTail`, we wouldn't have to re-identify yo and yi. They should be pointing to the ones in `x_inner` as before the split. The ones in `x_tail` should be the ones created new.

Reply:
In contrast to TVM we don't really have a "scheduling part" - our schedules, or transformations, are immediate and not postponed. This is one of the design principles we've been sticking so far. The main benefit of doing so is that everything becomes very explicit - i.e. we have to be very specific about what happens in every situation. It might seem laborious, but that's why we need higher-level transforms like "tile".

It's also worth noticing that postponing the part where we generate tail loops doesn't actually makes the solution simpler - it just moves the complexity to another part. For instance, if we apply `split_with_mask`-based tile for a simple loopnest:
```
for x in 0..10
  for y in 0..10
    a[x,y] = ...
```
we would get
```
for xo in 0..3
  for yo in 0..3
    for xi in 0..4
      if xo*4+xi < 10
        for yi in 0..4
          if yo*4+yi < 10
            a[xo*4+xi,yo*4+yi] = 
```
It is not straightforward how to convert these mask-based loops to tail-based loops, because they've been reordered already (e.g. the condition in the innermost loop depends on `yi` and `yo` and the latter is an index of a grand-parent loop).

I would actually claim that solving it after everything is done would be harder, because we would not have a context of how we ended up with such indexes there - the information that we had while we were inside the tile transformation.